src/SimpleMutex.php

```diff
@@ -0,0 +1,61 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Thesis\MessageBus\Transport\Amqp;
+
+use Amp\Cancellation;
+use Amp\Future;
+use function Amp\async;
+
+/**
+ * @api
+ * @template T
+ */
+final class SimpleMutex
+{
+    /**
+     * @var ?Future<T>
+     */
+    private ?Future $future = null;
+
+    /**
+     * @var ?T
+     */
+    private mixed $value = null;
+
+    /**
+     * @var \Closure(T): bool
+     */
+    private readonly \Closure $isHit;
+
+    /**
+     * @param \Closure(): T $factory
+     * @param ?\Closure(T): bool $isHit
+     */
+    public function __construct(
+        private readonly \Closure $factory,
+        ?\Closure $isHit = null,
+        private readonly ?Cancellation $cancellation = null,
```

Comment:
It would be better to pass `Cancellation` to the `get()` method, since the value might be unique to a specific request rather than the entire object where `OnceValue` is placed.